### DISCUSSION TITLE
How to expose APIs other than those automatically created for models

### DISCUSSION THREAD

**Discussion by Minal1409** [thomaxxl/safrs#104] [General]
I am using safrs for my project. I am not able to find how to expose api which is not a CRUD operation of any models. All/customised http methods are exposed for models. But we have couple of more APIs to be exposed.

TIA

**Comment by thomaxxl** [comment]
Hi,

Some info from the wiki:

https://github.com/thomaxxl/safrs/wiki/Instances-without-a-SQLAlchemy-model

https://github.com/thomaxxl/safrs/wiki/RPC

I will give a longer answer tomorrow, I am kinda busy atm.

**Reply by Minal1409** [reply]
Hi, Thanks a lot for prompt reply. 

I am able to expose APIs now.

But i am unable to understand few things.
1. How to define URL for exposed API
2. How URL is getting defined automatically by safrs json rpc decorator. It is taking query parameter in URL, which i have not even defined.
3. Does three attributes "description, args and parameters" are the way to provide description, request body and request query params respectively.

Thanks again

**Comment by thomaxxl** [comment]
Hi,

The rpc decorator url parameter can be omitted by decorating the method with `staticmethod` or `classmethod`  ([example](https://github.com/thomaxxl/safrs/blob/33cc252879799460e088145bf627a69405f8915d/examples/demo_pythonanywhere_com.py#L159))
I was going to give a longer reply but didn't find the time. Hopefully tomorrow.

**Comment by thomaxxl** [comment]
The generated API can be extended with custom routes using the flask `@app.route` decorator. If you want this to show up in the swagger, you will have to create the swagger/oas spec yourself and merge it with the generated oas ([example](https://github.com/thomaxxl/safrs/blob/33cc252879799460e088145bf627a69405f8915d/examples/demo_pythonanywhere_com.py#L269))

You can also use the safrs `@jsonapi_rpc` decorator. as explained in the wiki [here](https://github.com/thomaxxl/safrs/wiki/RPC) .

A third way to add custom endpoints is by simulating a safrs jsonapi class (https://github.com/thomaxxl/safrs/wiki/Instances-without-a-SQLAlchemy-model) . I noticed there's a bug here that will be fixed in the next release.
